include/zephyr/drivers/cellular.h

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (c) 2023 Bjarki Arge Andreasen
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file drivers/cellular.h
+ * @brief Public cellular network API
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_CELLULAR_H_
+#define ZEPHYR_INCLUDE_DRIVERS_CELLULAR_H_
+
+/**
+ * @brief Cellular interface
+ * @defgroup cellular_interface Cellular Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <zephyr/types.h>
+#include <zephyr/device.h>
+#include <errno.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define CELLULAR_MODEM_INFO_IMSI_SIZE 15
+#define CELLULAR_MODEM_INFO_IMEI_SIZE 16
+#define CELLULAR_MODEM_INFO_ICCID_SIZE 21
+#define CELLULAR_MODEM_INFO_MODEL_SIZE 16
+#define CELLULAR_MODEM_INFO_MANUFACTURER_SIZE 10
+
+/** Cellular access technologies */
+enum cellular_access_technology {
+	CELLULAR_ACCESS_TECHNOLOGY_GSM = 0,
+	CELLULAR_ACCESS_TECHNOLOGY_GPRS,
+	CELLULAR_ACCESS_TECHNOLOGY_UMTS,
+	CELLULAR_ACCESS_TECHNOLOGY_EDGE,
+	CELLULAR_ACCESS_TECHNOLOGY_LTE,
+	CELLULAR_ACCESS_TECHNOLOGY_LTE_CAT_M1,
+	CELLULAR_ACCESS_TECHNOLOGY_LTE_CAT_M2,
+	CELLULAR_ACCESS_TECHNOLOGY_NB_IOT,
+};
+
+/** Cellular network structure */
+struct cellular_network {
+	/** Cellular access technology */
+	enum cellular_access_technology technology;
+	/**
+	 * List of bands, as defined by the specified cellular access technology,
+	 * to enables. All supported bands are enabled if none are provided.
+	 */
+	uint16_t *bands;
+	/** Size of bands */
+	uint16_t size;
+};
+
+struct cellular_signal {
+	/* RSSI signal */
+	int16_t rssi;
+	/* Bit Error Rate */
+	uint8_t ber;
+};
```

Comment:
I don't think bit error rate should be used here as defined by the 3GPP specification at least, given that it is often excluded (set to 99) and is terribly complicated to decode
![image](https://github.com/zephyrproject-rtos/zephyr/assets/82143143/4d894be5-5a83-4c76-a6ab-442ea0950039)

We should find a good way to abstract this data to be useful to the user IMO, RSSI is quite useful and is AFAIK always available, but even this could be abstracted into a EXCELLENT, GOOD, BAD spectrum, or a percentage, like what has been attempted here https://usatcorp.com/faqs/understanding-lte-signal-strength-values/?__cf_chl_tk=S8.MOHvJvlqFnfSdCNbPHzD_fq9sPTvCtPyve0O8FzI-1700753956-0-gaNycGzNC1A#section-2-1

Reply:
do you agree about the concept of having a separate API for static / pull once `get_modem_info` and for dynamic /  pull several times `get_signal` ?

Reply:
I'm not sure, static modem info is quite limited right? IMEI and model number is pretty much the only non changing property I can think of which we have to poll from the modem, others depend on network selection, sim card inserted etc. Even properties like firmware versions can change during runtime if the modem is updated :)

I think overall the modem driver should determine how often in needs to read the static data, and the API should not care if the value is static or not, so I'm leaning towards a single API

Reply:
I'd like to suggest to add RSRP and RSRQ, the latter potentially replacing BER. RSRP is a better measurement of signal strength (relevant signal strength anyways), and commonly used in LTE.

Obviously if a modem is using GSM or 3G this might not be available, which should be indicated for example by setting RSRP/Q to INT16_MIN. But at least it should be there since 4G is the current and probably most used technology

Reply:
I'd like to suggest to add RSRP and RSRQ, the latter potentially replacing BER. RSRP is a better measurement of signal strength (relevant signal strength anyways), and commonly used in LTE.

Obviously if a modem is using GSM or 3G this might not be available, which should be indicated for example by setting RSRP/Q to INT16_MIN. But at least it should be there since 4G is the current and probably most used technology

Reply:
If all values are integers, we could create a getter like the one for modem_info, which allows us to define both RSSI, RSRP and RSRQ in an enum, and return -ENODATA if the value is not present. I really like to avoid reserved values if possible :)

Reply:
RSSI and RSRP are bith in dBm which is quite nice, RSRQ seems to be dB which makes it a bit less nice to include if we use a getter for signal strength

Reply:
> I'm not sure, static modem info is quite limited right? IMEI and model number is pretty much the only non changing property I can think of which we have to poll from the modem, others depend on network selection, sim card inserted etc. Even properties like firmware versions can change during runtime if the modem is updated :)
> 
> I think overall the modem driver should determine how often it needs to read the static data, and the API should not care if the value is static or not, so I'm leaning towards a single API

So far I also had in mind that there should be a separation between static (btw, I think these are IMEI, model number and manufacturer) and dynamic information. But @bjarki-trackunit has a good point. Things like the sim card and the fw version can change in runtime, so static information is very limited. It would be simpler to keep a single API for everything and in the end, it should be the application that decides what part of the data is static and what not.

On another note, is it actually the modem driver or the app that determines how often this info have to be read from the modem? Perhaps the time interval should be configurable?

Reply:
> I would expect the application to request RSSI etc. as often as it sees fit, I don't think we actually need to store it and poll periodically

Yes, that probably sounds the most rational thing to do.